### PR TITLE
First steps towards generalization in reimbursiments description

### PR DESCRIPTION
**Pull request by ghost**
It is the inclusion of https://github.com/datasciencebr/serenata-de-amor/pull/238 in Rosie.
Since the built classifier has 91% of accuracy, maybe it can be included in the set of classifiers for the chamber of deputies.

PS:
1) I uploaded the ML model (40 MB) together to code. I guess it is not the best practice, maybe it must be downloaded from the serenata-toolbox, however i don't have access to amazon to upload the model  :/ ( * Probably we have to change it in order to have a better architecture)

2) The code works fine in the Test:
screenshot from 2017-07-20 14-01-53
However i couldn't run it in the real reimbursements. Rosie is consuming all my RAM (5 Gb) while the reimbursements are downloaded, maybe someone could test it to me :+1: 

3) Sorry for my Python code. I just started to build something in Python there 4 months. I will appreciate some tips to improve the code.

4) As you can see in the print the TensorFlow i included is the basic one. I kept it since i don't know whether the machine where Rosie is running allows other configurations. (*We also can change it)

### CODE REVIEW

**Comment by luzfcb** [comment]
A comment about 40 MB of binary file (ML model)
Commit a binary file on git repository is a bad idea. I have problems with this decision, because in https://github.com/pythonclub/pythonclub.github.io we had commited too many images files, about 3MB in total. The result was that, over time, the size of the repository increases exponentially, even though I did not include any other new binary files. (So far, the pythonclub repository is close to 216.33MB) Removing these files forces me to rewrite a whole git history, and this would break all branchs and forks. So I can not do that.

In the last year, github has added support for Git LFS, which was apparently created to solve problems like this.

https://git-lfs.github.com/
https://help.github.com/articles/versioning-large-files/

File: Dockerfile [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
-  
-RUN pip install --upgrade pip
 
+RUN pip install --upgrade pip
 COPY requirements.txt ./
 COPY setup ./
 COPY rosie.py ./
 COPY rosie ./rosie
 COPY config.ini.example ./
+COPY config.ini ./
```
Comment (jtemporal) [inline Dockerfile]: This doesn't need to be here since Rosie's `setup` file creates `config.ini` from the example file copied in the previous line 😉 

You can go ahead and delete this line.

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+import pandas as pd
+from urllib.request import urlopen
+from sklearn.base import TransformerMixin
+from keras.preprocessing.image import ImageDataGenerator
+from keras.models import Sequential
+from keras.layers import Conv2D, MaxPooling2D
+from keras.layers import Activation, Dropout, Flatten, Dense
+from keras import backend as K
+from keras.callbacks import ModelCheckpoint
+from keras.models import load_model
```
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: You can follow [PEP-8 imports section](https://www.python.org/dev/peps/pep-0008/#imports) to guide you when organizing your imports. For example, when doing `from ... import ...` you can separate the second block with comma like: `from keras.models import Sequential, load_model`.

Note that you should also group your imports:
```Imports should be grouped in the following order:

1. standard library imports
1. related third party imports
1. local application/library specific imports

You should put a blank line between each group of imports.
```

Also on that note, you can make use of tools to help you automatically organize your imports like [isort](https://pypi.python.org/pypi/isort/).

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        the value is equal to "Meal".
+
+    document_id : string column
+        The identifier of the expense.
+
+    year : string column
+        The year the expense was generated.
+    """
+
+    COLS = ['applicant_id',
```
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: Don't be afraid to use descriptive names like `COLUMNS` for your variables. It will make your code clearer to read down the road.

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    def train(self,train_data_dir,validation_data_dir,save_dir):
+        #fix random seed for reproducibility
+        seed = 2017
+        np.random.seed(seed)
+
+        nb_train_samples = sum([len(files) for r, d, files in os.walk(train_data_dir)])
+        nb_validation_samples = sum([len(files) for r, d, files in os.walk(validation_data_dir)])
+
+        print('no. of trained samples = ', nb_train_samples, ' no. of validation samples= ',nb_validation_samples)
```
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: Long lines [aren't a good thing](https://www.python.org/dev/peps/pep-0008/#maximum-line-length). This print could be like the following in other to avoid the extra long line:
```
print('no. of trained samples = ', nb_train_samples,
      ' no. of validation samples= ', nb_validation_samples)
```
Extra space was missing there.

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+    COLS = ['applicant_id',
+            'document_id',
+            'category',
+            'year']
+
+
+    img_width, img_height = 300, 300
+
+    def train(self,train_data_dir,validation_data_dir,save_dir):
```
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: Pay extra attention to spaces after commas, they help make your code easier on the eyes 😉

This method would be nicer like this:
```
def train(self, train_data_dir, validation_data_dir, save_dir):
```

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        model.add(Activation('relu'))
+        model.add(MaxPooling2D(pool_size=(2, 2)))
+
+        model.add(Conv2D(32, (3, 3)))
+        model.add(Activation('relu'))
+        model.add(MaxPooling2D(pool_size=(2, 2)))
+
+        model.add(Conv2D(64, (3, 3)))
+        model.add(Activation('relu'))
+        model.add(MaxPooling2D(pool_size=(2, 2)))
```
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: I noted that you make the same process here 3 times with very little difference between parameters used. Maybe you can explain a little why is that necessary?

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+        model.add(Activation('relu'))
+        model.add(Dropout(0.5))
+        model.add(Dense(1))
+        model.add(Activation('sigmoid'))
+
+        model.compile(loss='binary_crossentropy',
+                      optimizer='rmsprop',
+                      metrics=['accuracy'])
+
+        #this is the augmentation configuration we will use for training
```
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: [PEP-8 inline comments](https://www.python.org/dev/peps/pep-0008/#inline-comments) state that:
>They (inline comments) should start with a # and a single space.

`# This is the augmentation configuration we will use for training` looks way better don't you think?

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        train_datagen = ImageDataGenerator(
+            rescale=1. / 255,
+            shear_range=0.2,
+            zoom_range=0.2,
+            horizontal_flip=False)#As you can see i put it as FALSE and on link example it is TRUE
```
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: [PEP-8 inline comments](https://www.python.org/dev/peps/pep-0008/#inline-comments) state that:
>An inline comment is a comment on the same line as a statement. Inline comments should be separated by at least two spaces from the statement. They should start with a # and a single space.

And be careful on the line length here too.

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        #Explanation, there no possibility to write in a reverse way :P
+
+        #this is the augmentation configuration we will use for testing:
```
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: Same thing as inline comments, these should start with `#` followed by a single space.

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 8 above)
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: Also, I didn't quite get what you meant here.

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+            class_mode='binary')
+
+        validation_generator = test_datagen.flow_from_directory(
+            validation_data_dir,
+            target_size=(img_width, img_height),
+            batch_size=batch_size,
+            class_mode='binary')
+
+        #It allow us to save only the best model between the iterations
+        checkpointer = ModelCheckpoint(filepath=save_dir+"weights.hdf5", verbose=1, save_best_only=True)
```
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: File paths should be built using `os.path.join` to avoid breaking the scripts when running Rosie in different systems

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 9 above)
Comment (jtemporal) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: This is applicable to all other comments you made in this file ;)

**Review by jtemporal — COMMENTED** [review]
First comments on 💅 only. PEP-8 for the win ;)

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 10 above)
Comment (luzfcb) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: @jtemporal Another good option with good API is [pathlib](https://docs.python.org/3/library/pathlib.html)

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 2 above)
Comment (silviodc) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: Done! Now they look beautiful :D

File: Dockerfile (hunk 1 above)
Comment (silviodc) [inline Dockerfile]: Done, i removed it. Strange the first time i tried to build the image i got error without it... Any way now its ok

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 3 above)
Comment (silviodc) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: In fact i tried to follow the code of other classifiers. Ctrl+C > Ctrl+V
I changed it to COLUMNS ;)

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 5 above)
Comment (silviodc) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: I changed the code using this tool: http://pep8online.com/checkresult
PEP8 online
Check your code for PEP8 requirements

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 6 above)
Comment (silviodc) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: I included a brief description and a link to explain it: http://deeplearning.net/tutorial/lenet.html

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 7 above)
Comment (silviodc) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: Done, using the tool

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 8 above)
Comment (silviodc) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: It was a copy and past from serenata. Now i included the reason

I put horizontal_flip as FALSE because we can not handwrite from right to left in Portuguese

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 10 above)
Comment (silviodc) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: Done, i included the os.path.join

File: rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py (hunk 9 above)
Comment (silviodc) [inline rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py]: PEP8 requirements done, using it: http://pep8online.com/checkresult

**Comment by silviodc** [comment]
Main points after the last pull request:

1.  I deleted the Machine Learn model  of 40MB 👍  Now i configured the code to load the model  from some informed path or download from external source (the one i'm using).
2. I included tests to predict downloaded reimbursements using the proper ML model 
3. I included a test to create and to train a fake model.
4. I included a test to this classifier in TestCore class
5. I fixed the location of supervised machine learn models in `rosie.chamber_of_deputies.settings ` and `rosie.federal_senate.settings`. Therefore, new supervised models will be able to specific the files they have to load.
